### PR TITLE
273 step chart

### PR DESCRIPTION
**Pull request by Jah25**
**Explain the details for making this change. What existing problem does the pull request solve?**

adding updated ids-step-chart component to the repository.

**Related github/jira issue (required)**:

closes [#273](https://github.com/infor-design/enterprise-wc/issues/273)

**Steps necessary to review your pull request (required)**:

- run the application locally
- navigate to http://localhost:4300/ids-step-chart/
- review the example cases
- run the functional test
- review requirements listed in [#273](https://github.com/infor-design/enterprise-wc/issues/273)

**Included in this Pull Request**:
- [X] An e2e or functional test for the bug or feature.
- [ ] A note to the change log.

### CODE REVIEW

File: src/components/ids-step-chart/README.MD [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+Step chart provides a custom element <ids-step-chart></ids-step-chart> that, when provided provides a stepped and labelled progress bar
+
+## Use Cases
+
+The step chart is designed to indicate progress through a series of steps, this can be a simple binary, step complete incomplete or optionally allows for a "in progress" state that indicates a step is partially completed. this also provides the option for labeling the chart with primary and secondary text indicators. see the features section for more details
+
+## Terminology
+
+Step: individual block in the component represented by a colored div. can be completed, in progress or untouched.
```
Comment (tmcconechy) [inline src/components/ids-step-chart/README.MD:13]:
```suggestion
**Step**: Individual block in the component represented by a colored `<div>` can be completed, in progress or not done.
```

File: src/components/ids-step-chart/README.MD [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+Completed: possible status of a step, will be displayed in a user selected color.
+
+In Progress: optional status of a step that has not been completed but touched, will be displayed in a user provided secondary color
```
Comment (tmcconechy) [inline src/components/ids-step-chart/README.MD:17]:
```suggestion
**In Progress**: An optional status of a step that has not been completed but started, will be displayed in a user provided secondary color.
```

File: src/components/ids-step-chart/README.MD [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+Untouched: dfault step status for step not in the other two catagories, is set by the component to a default graphite
```
Comment (tmcconechy) [inline src/components/ids-step-chart/README.MD:19]:
```suggestion
**Not Started**: Default step status for step not in the other two categories, is set by the component to a default graphite.
```

File: src/components/ids-step-chart/README.MD [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+## Features (With Code Examples)
+
+Ids step chart is created by using the <ids-step-chart> element.
+
```
Comment (tmcconechy) [inline src/components/ids-step-chart/README.MD:24]:
```suggestion
```

File: src/components/ids-step-chart/README.MD [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+for an implimentation with only a primary label and completed incompleted steps see the below example.
```
Comment (tmcconechy) [inline src/components/ids-step-chart/README.MD:25]:
```suggestion
For an implementation with only a primary label and completed uncompleted steps see the below example.
```

Not sure about this one? completed uncompleted?

File: src/components/ids-step-chart/README.MD [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+```html
+<ids-step-chart label="2 of 7 steps completed" color="azure06" step-number="7" value="2">
+</ids-step-chart>
+```
+
+to set in progress steps you must pass an array of strings or numbers where each item in the array is the number of the step you want to update and set the progress-color attribute
```
Comment (tmcconechy) [inline src/components/ids-step-chart/README.MD:32]:
```suggestion
To set in progress steps you must pass an array of strings or numbers where each item in the array is the number of the step you want to update and set the `progress-color` attribute.
```

File: src/components/ids-step-chart/README.MD [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+```html
+  <ids-step-chart label="2 of 7 steps completed" color="azure06" step-number="7" value="3" completed-label="5 days overdue" progress-color="ruby02">
+    <ids-icon id="alert" slot="icon" icon="alert"></ids-icon>
+  </ids-step-chart>
+```
+
+
+## Settings and Attributes
+
+  - color: the color used for completed steps (must use ids provided colors)
```
Comment (tmcconechy) [inline src/components/ids-step-chart/README.MD:58]:
```suggestion
  - `color` {string} Sets the color used for completed steps (must use ids provided colors).
```

Please add all these like:

```
- `color` {string} Sets the color of the badge e.g. `alert`, `error`, or `info`.
```

File: src/components/ids-step-chart/README.MD [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  - completed-label: secondary label for the component
+  - label: primary label for the compenent usually formatted as "x of y steps completed
+  - progress-color: color used for flaged steps in progress (must use ids provided colors)
+  - steps-in-progress: an array of steps that should be flagged as in progress
+  - step-number: total number of steps in the chart
+  - value: total number of steps completed or in progress
+
+## States and Variations
```
Comment (tmcconechy) [inline src/components/ids-step-chart/README.MD:66]: Can remove empty sections

File: src/components/ids-step-chart/README.MD [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+[design doc](https://www.figma.com/file/yaJ8mJrqRRej8oTsd6iT8P/IDS-(SoHo)-Component-Library-v4.5?node-id=760%3A771)
+
+## Converting from Previous Versions
+
+Add detailed but concise info on converting from previous version.
+
+## Accessibility Guidelines
+
+Keep in mind that individual steps are not focusable elements as they are purely visual. It is important to keep both primary and secondary labels up to date for screen reader users.
+
```
Comment (tmcconechy) [inline src/components/ids-step-chart/README.MD:85]: Extra spaces

File: src/components/ids-step-chart/README.MD [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+## Mobile Guidelines
+
+## Designs
+
+[design doc](https://www.figma.com/file/yaJ8mJrqRRej8oTsd6iT8P/IDS-(SoHo)-Component-Library-v4.5?node-id=760%3A771)
+
+## Converting from Previous Versions
+
+Add detailed but concise info on converting from previous version.
```
Comment (tmcconechy) [inline src/components/ids-step-chart/README.MD:80]: Could fill this in

File: src/components/ids-step-chart/ids-step-chart.d.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import { IdsElement } from '../../core';
+
+export default class IdsStepChart extends IdsElement {
+  color: string;
```
Comment (tmcconechy) [inline src/components/ids-step-chart/ids-step-chart.d.ts:4]: Add comments for these? See other examples

File: src/components/ids-step-chart/ids-step-chart.js [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+    this.#handleEvents();
+  }
+
+  /**
+   * Return the attributes we handle as getters/setters
+   * @returns {Array} The attributes in an array
+   */
+  static get attributes() {
+    return [
+
```
Comment (tmcconechy) [inline src/components/ids-step-chart/ids-step-chart.js:50]:
```suggestion
```

File: src/components/ids-step-chart/ids-step-chart.js [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+      attributes.LABEL,
+      attributes.PROGRESS_COLOR,
+      attributes.STEP_NUMBER,
+      attributes.VALUE
+    ];
+  }
+
+  get color() { return this.getAttribute(attributes.COLOR); }
+
+  set color(value) {
```
Comment (tmcconechy) [inline src/components/ids-step-chart/ids-step-chart.js:62]: Please add comments for all methods/setters. See other xamples

File: test/ids-step-chart/ids-step-chart-func-test.js [hunk 14]
```diff
@@ -0,0 +1,93 @@
+/**
+ * @jest-environment jsdom
+ */
+/* eslint-disable no-debugger */
```
Comment (tmcconechy) [inline test/ids-step-chart/ids-step-chart-func-test.js:4]: Might be able to remove this now?

File: test/ids-step-chart/ids-step-chart-func-test.js [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    expect(elem.shadowRoot.querySelector('.complete').getAttribute('color')).toBe('azure08');
+    elem.color = 'amethyst05';
+    expect(elem.shadowRoot.querySelector('.complete').getAttribute('color')).toBe('amethyst05');
+  });
+  it('can steps in progress be updated', () => {
+    elem.stepsInProgress = ['3', '5', '7'];
+    expect(elem.shadowRoot.querySelectorAll('.in-progress').length).toBe(3);
+    expect(elem.stepsInProgress).toEqual([3, 5, 7]);
+  });
+  it('completed label can be updated', () => {
```
Comment (tmcconechy) [inline test/ids-step-chart/ids-step-chart-func-test.js:75]: Add a space between each one?

File: test/ids-step-chart/ids-step-chart-func-test.js.snap [hunk 16]
```diff
@@ -0,0 +1,3 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`Ids Step Chart Tests renders correctly 1`] = `""`;
```
Comment (tmcconechy) [inline test/ids-step-chart/ids-step-chart-func-test.js.snap:3]: This is empty? Maybe use `.shadowRoot.innerHTML` in the snapshot test so it has something useful in it.

File: test/ids-step-chart/ids-step-chart-percy-test.js [hunk 17]
```diff
@@ -0,0 +1,26 @@
+import percySnapshot from '@percy/puppeteer';
+
+describe('Ids Tag Percy Tests', () => {
```
Comment (tmcconechy) [inline test/ids-step-chart/ids-step-chart-percy-test.js:3]:
```suggestion
describe('Ids Step Chart Percy Tests', () => {
```

**Review by tmcconechy — COMMENTED** [review]
@Jah25 looking through im not seeing that all the comments are addressed? Still working on it?

File: src/components/ids-step-chart/ids-step-chart.js [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+    }
+  }
+
+  /**
+   * Establish internal event handlers
+   * @private
+   * @returns {object} The object for chaining
+   */
+  #handleEvents() {
+    return this;
```
Comment (EdwardCoyle) [inline src/components/ids-step-chart/ids-step-chart.js:149]: I'm thinking that if there aren't any event handlers in this component, it's probably OK to take out `#handleEvents` and the code near the constructor that adds `IdsEventsMixin`.

**Review by EdwardCoyle — COMMENTED** [review]
@Jah25 Looks pretty good so far!  My comments are nitpicks too and will happily approve this after we address 👍

**Comment by tmcconechy** [comment]
Ok got it 👍🏻 . What you can do is tag the label WIP. Or just request review when its done.

Screen Shot 2021-09-01 at 3 25 19 PM

File: test/ids-step-chart/ids-step-chart-func-test.js (hunk 14 above)
Comment (Jah25) [reply test/ids-step-chart/ids-step-chart-func-test.js]: I found I was still having issues when I removed the jest environment comment here

Comment (tmcconechy) [reply test/ids-step-chart/ids-step-chart-func-test.js]: Sorry @Jah25  i just mean the `/* eslint-disable no-debugger */` part. As i didnt see any debugger statements in the code to ignore.

**Comment by Jah25** [comment]
@tmcconechy @EdwardCoyle just wanted to reach out about the Axe error's the build is getting. I'm not seeing them in a manual axe scan and the build doesn't say what the issue is. I was on a call with Rob earlier about it and we couldn't figure it out.

**Comment by tmcconechy** [comment]
I don't know why the build doesn't show it well but run it locale and you can see it with `npx jest -- step`

```
tmcconechy@usmvtmcconechy enterprise-wc2 % npx jest -- step
Dev server listening on port 4444
[percy] Percy is not running, disabling snapshots
 FAIL  test/ids-step-chart/ids-step-chart-e2e-test.js
  ● Ids Step Chart e2e Tests › should not have errors

    expect(received).resolves.toMatch(expected)

    Expected substring: "IDS Step Chart Component"
    Received string:    "Error"

       7 |
       8 |   it('should not have errors', async () => {
    >  9 |     await expect(page.title()).resolves.toMatch('IDS Step Chart Component');
         |                                         ^
      10 |   });
      11 |
      12 |   it('Should pass an Axe accessibility test', async () => {

      at Object.toMatch (node_modules/expect/build/index.js:242:22)
      at _callee2$ (test/ids-step-chart/ids-step-chart-e2e-test.js:9:41)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:63:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:293:22)
      at Generator.next (node_modules/regenerator-runtime/runtime.js:118:21)
      at asyncGeneratorStep (node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
      at _next (node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
      at node_modules/@babel/runtime/helpers/asyncToGenerator.js:32:7
      at Object.<anonymous> (node_modules/@babel/runtime/helpers/asyncToGenerator.js:21:12)

  ● Ids Step Chart e2e Tests › Should pass an Axe accessibility test

    expect(received).toPassAxeTests(expected)

    Expected page to pass Axe accessibility tests.
    Violations found:
    Rule: "landmark-one-main" (Document should have one main landmark)
    Help: https://dequeuniversity.com/rules/axe/4.3/landmark-one-main?application=axe-puppeteer
    Affected Nodes:
      html
        Fix ALL of the following:
          - Document does not have a main landmark.

    Rule: "page-has-heading-one" (Page should contain a level-one heading)
    Help: https://dequeuniversity.com/rules/axe/4.3/page-has-heading-one?application=axe-puppeteer
    Affected Nodes:
      html
        Fix ALL of the following:
          - Page must have a level-one heading.

    Rule: "region" (All page content should be contained by landmarks)
    Help: https://dequeuniversity.com/rules/axe/4.3/region?application=axe-puppeteer
    Affected Nodes:
      pre
        Fix ANY of the following:
        - Some page content is not contained by landmarks


      13 |     await page.setBypassCSP(true);
      14 |     await page.goto(url, { waitUntil: ['networkidle2', 'load'] });
    > 15 |     await expect(page).toPassAxeTests();
         |                        ^
      16 |   });
      17 | });
      18 |

      at _callee3$ (test/ids-step-chart/ids-step-chart-e2e-test.js:15:24)
      at tryCatch (node_modules/regenerator-runtime/runtime.js:63:40)
      at Generator.invoke [as _invoke] (node_modules/regenerator-runtime/runtime.js:293:22)
      at Generator.next (node_modules/regenerator-runtime/runtime.js:118:21)
      at asyncGeneratorStep (node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
      at _next (node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)

 PASS  test/ids-step-chart/ids-step-chart-func-test.js
 FAIL  test/ids-step-chart/ids-step-chart-percy-test.js
  ● Ids Step Chart Percy Tests › should not have visual regressions in new dark theme (percy)

    Evaluation failed: TypeError: Cannot read properties of null (reading 'setAttribute')

      at __puppeteer_evaluation_script__:2:59
      at ExecutionContext._evaluateInternal (node_modules/puppeteer/src/common/ExecutionContext.ts:273:13)
      at ExecutionContext.evaluate (node_modules/puppeteer/src/common/ExecutionContext.ts:140:12)

  ● Ids Step Chart Percy Tests › should not have visual regressions in new contrast theme (percy)

    Evaluation failed: TypeError: Cannot read properties of null (reading 'setAttribute')

      at __puppeteer_evaluation_script__:2:59
      at ExecutionContext._evaluateInternal (node_modules/puppeteer/src/common/ExecutionContext.ts:273:13)
      at ExecutionContext.evaluate (node_modules/puppeteer/src/common/ExecutionContext.ts:140:12)

Test Suites: 2 failed, 1 passed, 3 total
Tests:       4 failed, 11 passed, 15 total
Snapshots:   1 passed, 1 total
Time:        5.492 s, estimated 6 s
Ran all test suites matching /step/i.
```

So what is happening is the page is erroring and then the scan fails (because it doesnt load the page). Then i think the errors mess it up. So fix the error here and it should be good:

Screen Shot 2021-09-02 at 9 57 27 AM

**Comment by tmcconechy** [comment]
Try adding the events mixin.. Looks like its missing in the mix part

````
class IdsStepChart extends mix(IdsElement).with(IdsThemeMixin, IdsEventsMixin) {
```

**Comment by tmcconechy** [comment]
Yeah it's used in the IdsThemeMixin to handle the changing of themes event. I couldnt figure out how to make it include IdsEventsMixin at the time but i think its possible to do that now. cc @EdwardCoyle  on that?

**Comment by tmcconechy** [comment]
Is this how we do it @EdwardCoyle ?

https://github.com/infor-design/enterprise-wc/blob/main/src/mixins/ids-theme-mixin/ids-theme-mixin.js#L8
vs
https://github.com/infor-design/enterprise-wc/blob/main/src/mixins/ids-mask-mixin/ids-mask-mixin.js#L21

Of course you can just include IdsEventsMixin in the component as well but this might be better if it works?

**Comment by EdwardCoyle** [comment]
@tmcconechy @Jah25 Yes, you can follow the strategy in the Mask mixin for "depending on" other mixins.  If the Theme mixin needs it, ideally we include it as a dependency.

One thing to note though - we should try to figure out the effect this has on components that also include the Events Mixin directly (fx I worry that including the mixin "twice" on some components -- once directly, once via the Theme mixin -- could make for some kind of performance hit... just not sure how to test it.

**Comment by tmcconechy** [comment]
One way to test i guess is look in the output js for the component and see if it added it twice? Might be safer to leave it for now and just add back the mixin to the component

**Comment by Jah25** [comment]
So I just pushed a fix including the event mixin back into the component for now since that's a quick fix. should we create a work item for updating Themes and seeing what the impact is?

**Comment by tmcconechy** [comment]
Ok good call i made https://github.com/infor-design/enterprise-wc/issues/324

**Comment by Jah25** [comment]
I think everything should be good here though I can't figure out how to approve the Percy diff

**Comment by tmcconechy** [comment]
I’ll check it but ping me your information email and I’ll invite you so you can too

**Review by tmcconechy — APPROVED** [review]

**Review by tmcconechy — APPROVED** [review]

**Review by EdwardCoyle — APPROVED** [review]
Looks good!

**Pull request merged**